### DISCUSSION TITLE
Collection of Signal Patterns

### DISCUSSION THREAD

**Discussion by oravecz** [preactjs/signals#492] [Show and tell]
Probably not the most ideal spot for this, but as Signals explode in popularity across multiple frameworks, I can't seem to find a set of common patterns for using Signals. So, I thought we might use this discussion to collect them.

**Comment by oravecz** [comment]
### Debounce for Signal

```typescript
export const debounceSignal = <T>(
  targetSignal: Signal<T>,
  timeoutMs = 0,
): Signal<T> => {
  const debounceSignal = signal<T>(targetSignal.value);

  effect(() => {
    const value = targetSignal.value;
    const timeout = setTimeout(() => {
      debounceSignal.value = value;
    }, timeoutMs);
    return () => {
      clearTimeout(timeout);
    };
  });

  return debounceSignal;
};
```

**Comment by oravecz** [comment]
Here is a use case that came up today. We have a separate signal for each type of filter that drives the resource's fetcher.
<img width="708" alt="image" src="https://github.com/preactjs/signals/assets/297712/57600977-463c-49a6-bae7-e0cb8ff6fc5d">

As the user makes choices (e.g. clientNameSignal, clientStatusSignal, clientTypeSignal) that will affect the query parameters, we wanted a signal (clientCriteriaSignal) that collects these updated values, but doesn't notify its own subscribers until some sort of "trigger".

There are alternative ways to model this using booleans as triggers or duplicating some state in local variables, but I was wondering if there is a better pattern or whether this is a good use case to come up with a DeferredNotificationSignal or TriggerNotificationSignal that exposes a `notify()` method which will then notify subscribers, but otherwise holds its state.

**Comment by andywilsonuk** [comment]
I think you could achieve this using the existing signal functionality, something like:

```
const notify = signal(0)

const trigger = () => {
  // be a writer but not a reader
  notify.value = notify.peek() + 1
}

effect(() => {
  // subscribe to the signal
  if (notify.value > 0) {
    // do something
  }
})
```

It doesn't hold state like you mentioned however does have the advantage over boolean signals, in that there's no need to set the signal back to false once it has been processed.
